README.md
# Audio Stat

## ВВЕДЕНИЕ

1. Сия программа является свободным ПО под лицензией [GPL v3](https://www.gnu.org/licenses/gpl.html).
2. Программа создана и дорабатывается автором исключительно ради собственных
   нужд и развлечения, а также в соответствии с его представлениями об эргономике
   и функциональности.
3. Автор всех видал в гробу и ничего никому не должен, кроме явно
   прописанного в GPL.
4. Несмотря на вышеуказанное, автор совершенно не против, если программа
   подойдёт кому-то еще, кроме тех, под кем прогорит сиденье из-за пунктов
   с 1 по 3.

## НАЗНАЧЕНИЕ

Сбор статистики по параметрам аудиофайлов: частоте сэмплирования,
разрядности, битрейту и т.п.

## ЧТО ТРЕБУЕТ ДЛЯ РАБОТЫ

- Linux (или другую ОС, в которой заработает нижеперечисленное, напр.
  MS Windows с установленным MSYS2)
- Python 3.6 или новее
- GTK 3.20 или новее и соотв. модули gi.repository
- модуль mutagen для Python соотв. версии

**Внимание!** Работа ПО не под Linux не тестировалась и не гарантируется!


Changelog
1.2 ====================================================================
! изменён формат файла настроек, старые поля игнорируются
- убрана лишняя обвязка импорта модулей, которая всё равно не должна
  была работать нормально
+ в дереве файлов на странице статистики отображаются всплывающие
  подсказки с параметрами файлов
* отображение диапазонов параметров в дереве статистики сделано
  единообразным (была мешанина из "-" и "...")
- исправлены тексты меток на стартовой странице
+ добавлена фильтрация по метаданным аудиофайлов
! алгоритм фильтрации изменён и может быть изменён в следующих версиях
* список типов (расширений) файлов жёстко задан в программе,
  конкретные типы из списка можно выбирать в параметрах фильтрации
  на стартовой странице
+ учитываются три группы параметров аудиофайлов по разрядности и частоте
  сэмплирования:
    стандартного разрешения - 16 бит, 44.1-48 кГц,
    низкого - ниже 16/44.1,
    высокого - разрядность больше 16 бит И частота сэмплирования
    не меньше 44.1 кГц; файлы с разрядностью 16 бит и частотой выше 44.1
    кГц аудиофильским хайрезом не считаем, потому что это какой-то изврат;
+ добавлен учёт файлов с ошибками метаданных

1.1 ====================================================================
+ добавлена функция копирования полного пути выбранного на странице
  статистики файла или каталога в буфер обмена
+ прикручена кое-какая проверка на наличие внешних зависимостей
  (модулей, не входящих в стандартный комплект поставки Python)
  с выводом сообщения
- исправлены ошибки загрузки файла настроек
- исправлена ошибка, из-за которой в потрохах программы "терялся"
  путь к выбранному каталогу
- исправлено отображение страницы с прогрессбаром
* исправлен README.md

1.0 ====================================================================
* причёсаны потроха
+ добавлена суммарная статистика по частотам сэмплирования
+ добавлена суммарная статистика по разрядности

0.1 ====================================================================
! что-то работает, но не всё, и не всё так, как надо


ascommon.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

""" ascommon.py

    Copyright 2021 MC-6312

    his file is part of AudioStat.

    AudioStat is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    AudioStat is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with AudioStat.  If not, see <http://www.gnu.org/licenses/>."""


TITLE = 'AudioStat'
VERSION = '1.2'
TITLE_VERSION = '%s v%s' % (TITLE, VERSION)
COPYLEFT = '🄯 2021 MC-6312'
URL = 'https://github.com/mc6312/audiostat'


def set_from_str(v):
    """Преобразование строки, разделённой пробелами, во множество строк"""

    return set(map(lambda s: s.lower(), v.split(None)))


def set_to_str(v):
    return ' '.join(sorted(v))


def str_to_int(v, vmin, vmax):
    """Преобразование строки в целое с принудительным ограничением
    диапазона значений."""

    # если приехало что-то, непреобразуемое в целое -
    # падаем с исключением, его обработает что-то снаружи
    v = int(v)

    if v < vmin:
        v = vmin
    elif v > vmax:
        v = vmax

    return v


def str_to_bool(v):
    try:
        return bool(int(v))
    except ValueError:
        return v.lower() == 'true'


def disp_int_val_k(i):
    return '?' if not i else i if isinstance(i, str) else '%.1f' % (i / 1000.0)


def disp_int_val(i):
    return '?' if not i else str(i)


def disp_int_range(a, b):
    if a == b:
        return disp_int_val(a)
    else:
        return '%s - %s' % (disp_int_val(a), disp_int_val(b))


def disp_int_range_k(a, b):
    if a == b:
        return disp_int_val_k(a)
    else:
        return '%s - %s' % (disp_int_val_k(a), disp_int_val_k(b))


def disp_bool(b, vtrue):
    return None if not b else vtrue


def floor_ceil_int(v, vmin, vmax):
    if v < vmin:
        return vmin
    elif v > vmax:
        return vmax
    else:
        return v


class Representable():
    """Класс-костыль для упрощения написания метода __repr__()
    в классах с наследованием, для облегчения отладки классов.
    repr(obj) для экземпляра такого класса возвращает
    "прилизанную" строку, содержащую отображения только нужных,
    т.е. явно указанных в методе __repr_fields__() полей."""

    def __repr_fields__(self):
        """Костыль для наследования __repr__.
        Должен возвращать список, содержащий кортежи из двух
        элементов - названия поля и значения, при необходимости
        преобразованного в строку в виде, пригодном для __repr__;
        цельночисленные, булевские и т.п. значения следует
        возвращать "как есть"."""

        return []

    def __repr__(self):
        """Метод, вызываемый стандартной функцией repr().
        Вызывает метод __repr_fields__(), форматирует
        полученный от него список полей.
        Если список был пуст, пытается обработать self.__dict__
        и self.__slots__, если они есть.
        Возвращает отформатированную строку с именем класса и перечислением
        полей с их значениями."""

        def __rfld(name, value):
            if isinstance(value, str):
                s = '"%s"' % value
            elif isinstance(value, int) or isinstance(value, float) or isinstance(value, bool):
                s = str(value)
            else:
                s = repr(value)

            return '%s=%s' % (name, s)

        flds = self.__repr_fields__()

        if not flds:
            if hasattr(self, '__dict__'):
                flds = self.__dict__.items()
            elif hasattr(self, '__slots__'):
                flds = self.__slots__.items()

        return '%s(%s)' % (self.__class__.__name__,
            ', '.join(map(lambda f: __rfld(*f), flds)))


if __name__ == '__main__':
    print('[debugging %s]' % __file__)

    class Dummy(Representable):
        pass

    class DummyWithFields(Representable):
        def __init__(self):
            self.fieldName = 1

    class DummyBroken(Representable):
        def __repr_fields__(self):
            return []

        def __init__(self):
            self.fieldName = 1

    print(repr(Dummy()))
    print(repr(DummyWithFields()))
    print(repr(DummyBroken()))


asconfig.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-


""" asconfig.py

    Copyright 2021 MC-6312

    his file is part of AudioStat.

    AudioStat is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    AudioStat is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with AudioStat.  If not, see <http://www.gnu.org/licenses/>."""


import sys
import os.path
from configparser import ConfigParser

from audiostat import *
from ascommon import *


class Config(Representable):
    """Настройки AudioStat.

    Поля:
        lastDirectory:
            строка, путь к последнему просканированному каталогу;

        filterParams:
            экземпляр класса FilterParams."""

    __S_SETTINGS = 'settings'
    __V_LASTDIR = 'lastDirectory'

    __S_FILTERS = 'filters'

    def __init__(self):
        #
        # основные настройки
        #
        self.lastDirectory = os.path.expanduser('~')

        #
        # параметры фильтрации
        #
        self.filter = AudioFileFilter()

        #
        # подготовка к загрузке
        #

        #
        if sys.platform == 'linux':
            # ...а если дистрибутив не XDG-совместимый, то это не моя проблема
            self.pathConfig = os.path.expanduser('~/.config/audiostat.cfg')
        elif sys.platform == 'win32':
            # для винды и так сойдёт
            self.pathConfig = os.path.join(os.environ['USERPROFILE'], 'audiostat.cfg')
        else:
            # тут предполагаем что-то *nix-образное, пусть даже и макось
            self.pathConfig = os.path.expanduser('~/.audiostat.cfg')

    def load(self):
        if not os.path.exists(self.pathConfig):
            return

        # пытаемся загрузить конфиг
        cfg = ConfigParser()
        cfg.read(self.pathConfig)

        #
        # тащим значения из конфига в поля сего объекта
        #

        # основные
        self.lastDirectory = os.path.expanduser(cfg.get(self.__S_SETTINGS,
            self.__V_LASTDIR, fallback=self.lastDirectory))

        # фильтрация
        for pname in AudioFileFilter.PARAMETERS:
            s = cfg.get(self.__S_FILTERS, pname, fallback=None)

            if s is not None:
                try:
                    self.filter.set_parameter_str(pname, s)
                except Exception as ex:
                    raise ValueError('Invalid parameter "%s" in section "%s" of file "%s" -  %s' % (
                                     pname, self.__S_FILTERS, self.pathConfig, str(ex)))

    def save(self):
        cfg = ConfigParser()
        cfg.add_section(self.__S_SETTINGS)
        cfg.add_section(self.__S_FILTERS)

        # основные
        cfg.set(self.__S_SETTINGS, self.__V_LASTDIR, self.lastDirectory)

        # фильтрация
        for pname in AudioFileFilter.PARAMETERS:
            cfg.set(self.__S_FILTERS, pname, self.filter.get_parameter_str(pname))
        #
        with open(self.pathConfig, 'w+') as f:
            cfg.write(f)

    def __repr_fields__(self):
        return []


if __name__ == '__main__':
    print('[debugging %s]' % __file__)

    cfg = Config()
    print('\033[1mdefaults:\033[0m\n', cfg)

    cfg.load()
    print('\033[1mloaded:\033[0m\n', cfg)

    cfg.save()


audiostat.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

""" audiostat.py

    Copyright 2021 MC-6312

    his file is part of AudioStat.

    AudioStat is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    AudioStat is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with AudioStat.  If not, see <http://www.gnu.org/licenses/>."""


from ascommon import *


import sys
import os
import os.path
import mutagen
from collections import namedtuple, OrderedDict
from enum import IntEnum
from traceback import print_exception


__aft = namedtuple('__aft', 'name exts')

AUDIO_FILE_TYPES = (
    __aft('FLAC',            {'.flac'}),
    __aft('WavPack',         {'.wv'}),
    __aft('Monkey’s Audio',  {'.ape'}),
    __aft('MPEG Layer 3',    {'.mp3'}),
    __aft('MPEG4 Audio',     {'.m4a'}),
    __aft('OGG Vorbis',      {'.ogg', '.oga'}),
    __aft('OGG Opus',        {'.opus'}),
    __aft('OptimFROG',       {'.ofr'}),
    __aft('AIFF',            {'.aif', '.aiff', '.aifc'}),
    __aft('Wave',            {'.wav'}),
    # форматы, (в том числе) не поддерживаемые mutagen,
    # но могущие оказаться в аудиотеке
    __aft('Other formats',   {'.webm'}),
    )


DEFAULT_AUDIO_FILE_EXTS = set()

for __, __exts in AUDIO_FILE_TYPES:
    DEFAULT_AUDIO_FILE_EXTS.update(__exts)


#TODO пополнить LOSSLESS_MIMETYPES при необходимости
LOSSLESS_MIMETYPES = {'audio/flac', 'audio/x-ape', 'audio/x-wavpack'}


TAGS = (('title', ('TITLE', 'TIT2')),
        ('artist', ('ARTIST', 'TPE1')),
        ('album artist', ('ALBUMARTIST', 'TPE2')),
        ('album', ('ALBUM', 'TALB')),
        ('track number', ('TRACKNUMBER', 'TRCK')),
        ('genre', ('TCON', 'GENRE')),
        ('year', ('DATE', 'TDRC')),
        )


# значение порога для фильтрации
DEFAULT_MIN_BITRATE = 192


class BaseAudioInfo(Representable):
    def get_info_strings(self):
        """Возвращает список строк для форматирования
        текста с информацией об объекте."""

        return []

    def get_hint_str(self):
        return '\n'.join(self.get_info_strings())


class AudioStreamInfo(BaseAudioInfo):
    RESOLUTION_LOW, RESOLUTION_STANDARD, RESOLUTION_HIGH = range(3)
    RESOLUTION_MIN = RESOLUTION_LOW
    RESOLUTION_MAX = RESOLUTION_HIGH

    BITRATE_MIN = 8
    BITRATE_MAX = 1000000 #!

    """Информация об аудиопотоке:
    lossy           - булевское; True, если использовано сжатие с потерями;
    resolution      - None или RESOLUTION_*; "разрешение" потока по значениям
                      sampleRate и bitsPerSample;
                      параметры проверки см. в функции get_audio_file_info();
    sampleRate      - целое, частота сэмплирования,
    channels        - целое, кол-во каналов;
    bitsPerSample   - целое, разрядность; м.б. 0 (неизвестно) для MP3 и
                      подобных форматов;
    bitRate         - целое, битрейт в килобитах/сек. (для форматов, где он известен);
    missingTags     - целое, битовые флаги TAG_xxx; ненулевое значение
                      в случае отсутствия важных тэгов в метаданных файла."""

    def __init__(self):
        self.reset()

    def reset(self):
        self.lossy = True
        self.resolution = None
        self.sampleRate = 0
        self.channels = 0
        self.bitsPerSample = 0
        self.bitRate = 0
        self.missingTags = 0

    def get_info_strings(self):
        r = super().get_info_strings()

        mt = missing_tags_to_str(self.missingTags)
        if mt:
            r.append('Missing tags: %s' % mt)

        return r


def missing_tags_to_str(mtflags):
    """Возвращает строку со списком тэгов,
    соответствующих битовым полям mtflags (целого)."""

    mt = []

    for ixtag, tag in enumerate(TAGS):
        if mtflags & (1 << ixtag):
            mt.append(tag[0])

    return  ', '.join(mt)


class AudioFileInfo(AudioStreamInfo):
    """Информация об аудиофайле.

    Поля:
        error   - None или строка с сообщением об ошибке,
                  если произошла ошибка разбора метаданных;
                  в этом случае все прочие поля должны
                  игнорироваться;
        mime    - строка, mimetype;

    Прочие поля наследуются от AudioStreamInfo."""

    def __init__(self):
        super().__init__()

        self.error = None
        self.mime = ''

    def get_info_strings(self):
        r = super().get_info_strings()

        if self.error:
            r.append('Error: %s' % self.error)

        return r


class AudioDirectoryInfo(BaseAudioInfo):
    """Класс для сбора статистики по каталогу с аудиофайлами.

    minInfo, maxInfo - экземпляры AudioStreamInfo,
    содержащие соответствующие значения после одного
    или более вызовов метода update_from_file();

    nFiles - целое, количество обработанных файлов."""

    def __init__(self):
        self.nFiles = 0
        self.nErrors = 0
        self.minInfo = AudioStreamInfo()
        self.maxInfo = AudioStreamInfo()

        self.reset()

    def reset(self):
        """Сброс счётчиков"""

        self.nFiles = 0
        self.nErrors = 0

        self.minInfo.reset()
        # в "минимальное" поле кладём максимальные допустимые значения!
        self.minInfo.lossy = False
        self.minInfo.resolution = AudioStreamInfo.RESOLUTION_MAX

        self.minInfo.sampleRate = 100000000
        self.minInfo.channels = 16384
        self.minInfo.bitsPerSample = 1024
        self.minInfo.bitRate = 1000000000

        self.maxInfo.reset()
        self.maxInfo.resolution = AudioStreamInfo.RESOLUTION_MIN

    def flush(self):
        """Сброс неиспользованных счётчиков.
        Метод должен вызываться после завершения обхода каталога."""

        if self.minInfo.sampleRate > self.maxInfo.sampleRate:
            self.minInfo.sampleRate = 0

        if self.minInfo.channels > self.maxInfo.channels:
            self.minInfo.channels = 0

        if self.minInfo.bitsPerSample > self.maxInfo.bitsPerSample:
            self.minInfo.bitsPerSample = 0

        if self.minInfo.bitRate > self.maxInfo.bitRate:
            self.minInfo.bitRate = 0

        if self.minInfo.resolution is not None\
           and self.maxInfo.resolution is not None\
           and self.minInfo.resolution > self.maxInfo.resolution:
            self.minInfo.resolution = None

    def __update_min(self, nfo):
        """Пополнение "минимальных" счётчиков из экземпляра
        AudioStreamInfo."""

        if self.minInfo.sampleRate > nfo.sampleRate:
            self.minInfo.sampleRate = nfo.sampleRate

        if self.minInfo.channels > nfo.channels:
            self.minInfo.channels = nfo.channels

        if self.minInfo.bitsPerSample > nfo.bitsPerSample:
            self.minInfo.bitsPerSample = nfo.bitsPerSample

        if self.minInfo.bitRate > nfo.bitRate:
            self.minInfo.bitRate = nfo.bitRate

        if nfo.resolution is not None:
            if self.minInfo.resolution < nfo.resolution:
                self.minInfo.resolution = nfo.resolution

    def __update_max(self, nfo):
        """Пополнение "максимальных" счётчиков из экземпляра
        AudioStreamInfo."""

        if self.maxInfo.sampleRate < nfo.sampleRate:
            self.maxInfo.sampleRate = nfo.sampleRate

        if self.maxInfo.channels < nfo.channels:
            self.maxInfo.channels = nfo.channels

        if self.maxInfo.bitsPerSample < nfo.bitsPerSample:
            self.maxInfo.bitsPerSample = nfo.bitsPerSample

        if self.maxInfo.bitRate < nfo.bitRate:
            self.maxInfo.bitRate = nfo.bitRate

        if nfo.resolution is not None:
            if self.maxInfo.resolution > nfo.resolution:
                self.maxInfo.resolution = nfo.resolution

    def update_from_dir(self, other):
        """Пополнение статистики из другого экземпляра
        (напр. при рекурсивном обходе каталогов)."""

        self.nFiles += other.nFiles
        self.nErrors += other.nErrors

        self.__update_min(other.minInfo)
        self.__update_max(other.maxInfo)

        self.minInfo.lossy |= other.minInfo.lossy
        self.minInfo.missingTags |= other.minInfo.missingTags

    def update_from_file(self, nfo):
        """Пополнение статистики.
        nfo - экземпляр AudioFileInfo."""

        self.nFiles += 1

        if nfo.error:
            self.nErrors += 1
        else:
            self.__update_min(nfo)
            self.__update_max(nfo)

            self.minInfo.lossy |= nfo.lossy

            self.minInfo.missingTags |= nfo.missingTags

    def get_info_strings(self):
        r = super().get_info_strings()

        mt = missing_tags_to_str(self.minInfo.missingTags)
        if mt:
            r.append('Missing tags: %s' % mt)

        if self.nErrors:
            r.append('Invalid files: %d' % self.nErrors)

        return r


class AudioFileFilter(Representable):
    """Параметры фильтрации аудиофайлов.

    Поля:
        byFileTypes:
            булевское, True - фильтровать по типам (расширениям) файлов;
        fileTypes:
            множество строк - типов (расширений) файлов;
            по умолчанию - DEFAULT_AUDIO_FILE_EXTS;

        byContainsStreamParameters:
            булевское, True - фильтровать по наличию параметров
            аудиопотока в файле;
        onlyContainsStreamParameters:
            булевское, True - учитывать только файлы, содержащие
            параметры аудиопотока;
            этот фильтр предназначен для выявления файлов, не поддерживаемых
            mutagen (т.е. с ненулевой вероятностью - экзотических
            или устаревших форматов);

        byLossless:
            булевское, True - фильтровать по формату сжатия (без потерь/
            с потерями);
        onlyLossless:
            булевское, True - учитывать только аудио, сжатое без потерь;

        byResolution:
            булевское, True - фильтровать по разрешению;
        resolution:
            AudioStream.Resolution;

        byBitrate:
            булевское, True - фильтровать по битрейту;
        bitrateLowerThan:
            булевское, True - учитывать аудио с битрейтом ниже указанного
            значения;
        bitrateLowerThanValue:
            целое, максимальное значение битрейта (используется, если
            BitrateLowerThan==True);
        bitrateGreaterThanValue:
            целое, минимальное значение битрейта (используется, если
            BitrateLowerThan==False);

        byTags:
            булевское, True - фильтровать по наличию всех важных тэгов
            (перечисленных в списке TAGS);
        onlyWithoutTags:
            булевское, True - показывать только файлы, где нет хотя бы
            одного важного тэга;

        byErrors:
            булевское, True - фильтровать файлы по наличию ошибок
            обработки (разбора mutagen'ом);
        withErrors:
            булевское, True - показывать только файлы с ошибками."""

    __fpar = namedtuple('__fpar', 'defval tostr fromstr')

    # 'name':(default, tostr, fromstr)
    PARAMETERS = OrderedDict({
        'byFileTypes': __fpar(False, str, str_to_bool),
        'fileTypes': __fpar(DEFAULT_AUDIO_FILE_EXTS, set_to_str, set_from_str),
        'byContainsStreamParameters': __fpar(False, str, str_to_bool),
        'onlyContainsStreamParameters': __fpar(False, str, str_to_bool),
        'byMissingTags': __fpar(False, str, str_to_bool),
        'onlyMissingTags': __fpar(True, str, str_to_bool),
        'byLossless': __fpar(False, str, str_to_bool),
        'onlyLossless': __fpar(True, str, str_to_bool),
        'byResolution': __fpar(False, str, str_to_bool),
        'resolution': __fpar(AudioStreamInfo.RESOLUTION_LOW, str,
                            lambda s: str_to_int(s,
                                AudioStreamInfo.RESOLUTION_MIN,
                                AudioStreamInfo.RESOLUTION_MAX)),
        'byBitrate': __fpar(False, str, str_to_bool),
        'bitrateLowerThan': __fpar(True, str, str_to_bool),
        'bitrateLowerThanValue': __fpar(DEFAULT_MIN_BITRATE, str,
                            lambda s: str_to_int(s,
                                AudioStreamInfo.BITRATE_MIN,
                                AudioStreamInfo.BITRATE_MAX)),
        'bitrateGreaterThanValue': __fpar(DEFAULT_MIN_BITRATE, str,
                            lambda s: str_to_int(s,
                                AudioStreamInfo.BITRATE_MIN,
                                AudioStreamInfo.BITRATE_MAX)),
        'byErrors': __fpar(False, str, str_to_bool),
        'onlyWithErrors': __fpar(False, str, str_to_bool),
        })

    def __init__(self):
        for pname, fpar in self.PARAMETERS.items():
            setattr(self, pname, fpar.defval)

    def get_parameter_str(self, pname):
        """Возвращает значение параметра с именем pname, преобразованное
        в строку."""

        # ЗДЕСЬ: если pname неправильное, просто рухнем с исключением
        # проверкой имён занимается класс Config
        fpar = self.PARAMETERS[pname]

        return fpar.tostr(getattr(self, pname))

    def set_parameter_str(self, pname, v):
        """Устанавливает значение параметра с именем "pname",
        преобразовав его в нужный тип из строки."""

        # ЗДЕСЬ: если pname неправильное, просто рухнем с исключением
        # проверкой имён занимается класс Config
        fpar = self.PARAMETERS[pname]

        try:
            setattr(self, pname, fpar.fromstr(v))
        except Exception as ex:
            raise ValueError('%s.set_parameter_str(): invalid value of parameter "%s" (%s)' % (
                self.__class__.__name__, pname, str(ex)))

    def filetypes_from_str(self, fts):
        self.fileTypes = set_from_str(fts.split(None))

    def filetypes_to_str(self):
        return set_to_str(self.fileTypes)

    def get_audio_file_info(self, fpath):
        """Проверка типа файла и извлечение параметров потока
        и метаданных из аудиофайла.

        Параметры:
            fpath   - строка, полный путь к файлу;
            fexts   - множество строк - расширений (типов) файлов.

        Возвращает экземпляр AudioFileInfo, если файл - поддерживаемого
        типа и соответствует параметрам фильтрации,
        в прочих случаях - None."""

        def __get_info_fld(info, name, fallback):
            if name in info.__dict__:
                return getattr(info, name)
            else:
                return fallback

        fext = os.path.splitext(fpath)[-1].lower()

        # расширение проверяем в любом случае:
        # если указано "проверять тип" - по выбранным типам
        # иначе - по всем известным типам
        if fext not in (self.fileTypes if self.byFileTypes else DEFAULT_AUDIO_FILE_EXTS):
            return

        nfo = AudioFileInfo()

        def __has_tags(fnfo, tnames):
            for n in tnames:
                if n in fnfo:
                    return True

            return False

        try:
            f = mutagen.File(fpath)

            #
            # фильтрация по указанным параметрам.
            # файлы, не прошедшие фильтрацию - отбрасываем
            #

            if self.byErrors and self.onlyWithErrors:
                # файл без ошибок, а тут мы хотим одних лишь ошибок
                return

            #
            hasParameters = 0

            if f:
                # извлекаем параметры и метаданные,
                # фильтровать по всему этому будем потом

                #
                nfo.mime = str(f.mime[0])
                nfo.lossy = nfo.mime not in LOSSLESS_MIMETYPES

                #
                nfo.sampleRate = __get_info_fld(f.info, 'sample_rate', 0)
                nfo.channels = __get_info_fld(f.info, 'channels', 1)
                nfo.bitsPerSample = __get_info_fld(f.info, 'bits_per_sample', 0)
                nfo.bitRate = int(__get_info_fld(f.info, 'bitrate', 0) / 1024)

                for v in (nfo.sampleRate, nfo.channels, nfo.bitsPerSample, nfo.bitRate):
                    if v > 0:
                        hasParameters += 1

                #
                tags = getattr(f, 'tags', None)
                if tags:
                    nfo.missingTags = 0

                    for ix, (_, tnames) in enumerate(TAGS):
                        if not __has_tags(f, tnames):
                            nfo.missingTags = nfo.missingTags or (1 << ix)
                # для тэгов hasParameters не трогаем, тэги - не параметры аудиопотока

            #
            # теперь уже пытаемся фильтровать
            #

            #
            if self.byContainsStreamParameters:
                if self.onlyContainsStreamParameters:
                    if not hasParameters:
                        return
                else:
                    if hasParameters:
                        return

            #
            if self.byLossless and self.onlyLossless and nfo.lossy:
                return

            #
            # пока проверка "на хайрез" приколочена гвоздями здесь
            #
            # ВНИМАНИЕ! файлы
            if nfo.bitsPerSample < 16 or nfo.sampleRate < 44100:
                nfo.resolution = AudioStreamInfo.RESOLUTION_LOW
            elif nfo.bitsPerSample > 16 and nfo.sampleRate >= 44100:
                nfo.resolution = AudioStreamInfo.RESOLUTION_HIGH
            else:
                nfo.resolution = AudioStreamInfo.RESOLUTION_STANDARD

            if self.byResolution and self.resolution != nfo.resolution:
                return

            #
            if self.byBitrate:
                if self.bitrateLowerThan:
                    if nfo.bitRate > self.bitrateLowerThanValue:
                        return
                elif nfo.bitRate < self.bitrateGreaterThanValue:
                    return

            #

            if self.byMissingTags:
                if (self.onlyMissingTags and nfo.missingTags == 0) or\
                        (not self.onlyMissingTags and nfo.missingTags != 0):
                    return

        except mutagen.MutagenError as ex:
            # с прочими исключениями - обязательно падаем!

            if not self.byErrors:
                return

            nfo.error = str(ex)

        return nfo


def __test_scan_directory(path, cfg):
    print('\033[1m%s/\033[0m' % path)

    for fname in os.listdir(path):
        fpath = os.path.join(path, fname)

        if os.path.isdir(fpath):
            __test_scan_directory(fpath, cfg)
        else:
            print('\033[32m%s\033[0m' % fname)
            nfo = cfg.filter.get_audio_file_info(fpath)
            if nfo:
                if not nfo.error:
                    print(nfo)
                else:
                    print('\033[31m*** Error: %s\033[0m' % nfo.error)


if __name__ == '__main__':
    print('[debugging %s]' % __file__)

    import asconfig

    cfg = asconfig.Config()
    cfg.load()
    print(cfg)
    __test_scan_directory(cfg.lastDirectory, cfg)
